File: apps/api/app/models/task.py
import uuid

from sqlalchemy import Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.sql import func

from app.database import Base


def _task_id():
    return f"task_{uuid.uuid4().hex[:8]}"


def _assign_id():
    return f"asgn_{uuid.uuid4().hex[:8]}"


class Task(Base):
    __tablename__ = "tasks"

    id = Column(String, primary_key=True, index=True, default=_task_id)
    room_id = Column(String, ForeignKey("rooms.id", ondelete="CASCADE"), nullable=False, index=True)
    title = Column(String, nullable=False)
    description = Column(Text, default="")
    prompt = Column(Text, default="")  # AI prompt for task execution
    status = Column(String, default="pending")  # pending, assigned, in_progress, completed, failed
    priority = Column(Integer, default=0)  # 0=normal, 1=high, 2=urgent
    created_by = Column(String, nullable=True)  # member_id of creator
    created_at = Column(DateTime(timezone=True), server_default=func.now())


class Assignment(Base):
    __tablename__ = "assignments"

    id = Column(String, primary_key=True, index=True, default=_assign_id)
    task_id = Column(String, ForeignKey("tasks.id", ondelete="CASCADE"), nullable=False, index=True)
    member_id = Column(String, ForeignKey("members.id", ondelete="CASCADE"), nullable=False, index=True)
    job_type = Column(String, nullable=False)  # one of the 25 job definition IDs
    status = Column(String, default="assigned")  # assigned, in_progress, completed, failed
    started_at = Column(DateTime, nullable=True)
    completed_at = Column(DateTime, nullable=True)
    result = Column(Text, default="{}")  # JSON — output data from the assignment

File: packages/shared-types/src/task.ts
/**
 * Task and Assignment shared types.
 * Aligned with: apps/api/app/models/task.py, apps/desktop/src/store/useRoomStore.ts
 */

// ── Task Status ──

export type TaskStatus =
  | "pending"
  | "assigned"
  | "in_progress"
  | "completed"
  | "failed";

export type TaskPriority = 0 | 1 | 2;

export const TASK_PRIORITY_LABELS: Record<TaskPriority, string> = {
  0: "Normal",
  1: "High",
  2: "Urgent",
};

// ── Task ──

export interface Task {
  id: string;
  room_id: string;
  title: string;
  description: string;
  prompt: string;
  status: TaskStatus;
  priority: TaskPriority;
  created_by: string | null;
  created_at: string;
}

/** Frontend-enriched task with display info. */
export interface RoomTask {
  id: string;
  title: string;
  prompt?: string;
  status: TaskStatus;
  assigneeId?: string;
  assigneeName?: string;
  createdAt: string;
  completedAt?: string;
}

// ── Assignment ──

export type AssignmentStatus =
  | "assigned"
  | "in_progress"
  | "completed"
  | "failed";

export interface Assignment {
  id: string;
  task_id: string;
  member_id: string;
  job_type: string;
  status: AssignmentStatus;
  started_at: string | null;
  completed_at: string | null;
  result: Record<string, unknown>;
}

// ── DTOs ──

export interface TaskCreate {
  title: string;
  description?: string;
  prompt?: string;
  priority?: TaskPriority;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  prompt?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
}
